### PR TITLE
ci: Adding e2e-tests for extra capabilities

### PR DESCRIPTION
**Pull request by devotional-phoenix-97**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [x] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

We don;t have unit tests for new capabilities

Fixes: # (issue)

## What is the new behavior?

- cap5 is also added to config
- when a paas with cap5 is created all works as expected
- when cap6 (not defined in COnfig) is created an error is reported back in Paas status

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: internal/controller/cluster_wide_quota.go [hunk 1]
```diff
@@ -207,7 +207,8 @@ func (r *PaasReconciler) removeFromClusterWideQuota(ctx context.Context, paas *v
 	var capConfig v1alpha1.ConfigCapability
 	var exists bool
 	if capConfig, exists = GetConfig().Capabilities[capabilityName]; !exists {
-		return fmt.Errorf("capability %s does not seem to exist", quotaName)
+		//return fmt.Errorf("capability %s does not seem to exist", quotaName)
+		return nil
```
Comment (portly-halicore-76) [inline internal/controller/cluster_wide_quota.go:212]: Alright, agreed! This must not return an error, else that would block the Paas from being finalized successfully. Please state that as a comment in the code for our next generation and remove the now commented out line.

File: internal/controller/paas_controller.go
Comment (portly-halicore-76) [inline internal/controller/paas_controller.go:1]: Great catch, forget those in my prev PR. Thanks

File: test/e2e/paas.go [hunk 2]
```diff
@@ -73,3 +81,27 @@ func waitForPaasReconciliation(ctx context.Context, cfg *envconf.Config, paas *a
 
 	return nil
 }
+
+// waitForPaasReconciliation polls a Paas resource, blocking until the Paas status reports successful reconciliation.
+func waitForPaasErrorReconciliation(ctx context.Context, cfg *envconf.Config, paas *api.Paas, msg string, oldGeneration int64) error {
```
Comment (portly-halicore-76) [inline test/e2e/paas.go:84]: Rather nice!! Need this somewhere else as well. Great solution.

File: test/e2e/steps.go [hunk 3]
```diff
@@ -21,7 +21,23 @@ func createPaasFn(name string, paasSpec api.PaasSpec) types.StepFunc {
...
-		if err := createPaasSync(ctx, cfg, paas); err != nil {
+		if err := createPaasSyncSuccess(ctx, cfg, paas); err != nil {
+			t.Fatal(err)
+		}
+
+		return ctx
+	}
+}
+
+// createPaasFn accepts a Paas spec object and a name and creates the Paas resource.
```
Comment (portly-halicore-76) [inline test/e2e/steps.go]: Pls update the doc accordingly

File: test/e2e/paas.go [hunk 4]
```diff
@@ -51,7 +59,7 @@ func deletePaasSync(ctx context.Context, name string, t *testing.T, cfg *envconf
 }
 
 // waitForPaasReconciliation polls a Paas resource, blocking until the Paas status reports successful reconciliation.
```
Comment (portly-halicore-76) [inline test/e2e/paas.go]: Pls update the doc accordingly

File: test/e2e/capability-cap6_test.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	"sigs.k8s.io/e2e-framework/pkg/features"
+)
+
+const (
+	paasWithCapability6 = "cap6paas"
+	cap6Name            = "cap6"
+	paasCap6Ns          = "cap6paas-cap6"
+	paasCap6CRQ         = "paas-cap6"
+	cap6ApplicationSet  = "cap6as"
+	cap6StatusMessage   = "Capability 'cap6' is not configured"
```
Comment (portly-halicore-76) [inline test/e2e/capability-cap6_test.go]: Pls remove this unused line.

File: test/e2e/capability-cap6_test.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+		features.New("Capability 6").
+			Setup(createPaasWithErrorFn(paasWithCapability6, paasSpec, "a capability is requested, but not configured")).
+			Assess("is created", assertCap6PaasCreated).
+			Assess("is deleted when PaaS is deleted", assertCap6Deleted).
+			Teardown(teardownPaasFn(paasWithCapability6)).
+			Feature(),
+	)
+}
+
+func assertCap6PaasCreated(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
```
Comment (portly-halicore-76) [inline test/e2e/capability-cap6_test.go]: I understand the point of this assertion by examining the rest of the code. However, both the name of this func, and the name of the `getAndFail` func can be improved stating something about making sure that none of these resources are created. You get the point.

File: test/e2e/capability-cap6_test.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	getAndFail(ctx, paasCap6Ns, cfg.Namespace(), &corev1.Namespace{}, t, cfg)
+	getAndFail(ctx, paasCap6CRQ, cfg.Namespace(), &quotav1.ClusterResourceQuota{}, t, cfg)
+
+	// Paas has correct name
+	assert.Equal(t, paasWithCapability6, paas.Name)
+
+	return ctx
+}
+
+func assertCap6Deleted(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
```
Comment (portly-halicore-76) [inline test/e2e/capability-cap6_test.go:58]: I have my doubt whether this adds value. In `assertCap6PaasCreated` we already asserted this quota had not been created. The assertion whether the namespace is removed after Paas has been removed is tested in the namespace related e2e test. Might remove this assert?

**Review by portly-halicore-76 — CHANGES_REQUESTED** [review]
Great addition to the test suite. Some minor improvements, see comments.

File: internal/controller/paas_controller.go [hunk 8]
```diff
@@ -194,23 +193,63 @@ func (r *PaasReconciler) Reconcile(ctx context.Context, req ctrl.Request) (resul
 	paas.Status.Truncate()
 
 	if err := r.ReconcileQuotas(ctx, paas); err != nil {
-		return ctrl.Result{}, errors.Join(err, r.setErrorCondition(ctx, paas, err))
+		err = r.setErrorCondition(ctx, paas, err)
```
Comment (portly-halicore-76) [inline internal/controller/paas_controller.go]: This is old. Please rebase on upstream.

**Review by portly-halicore-76 — DISMISSED** [review]

File: internal/controller/cluster_wide_quota.go (hunk 1 above)
Comment (devotional-phoenix-97) [reply internal/controller/cluster_wide_quota.go:212]: Comment added

File: test/e2e/capability-cap6_test.go (hunk 6 above)
Comment (devotional-phoenix-97) [reply test/e2e/capability-cap6_test.go]: I agree.

- renamed `getAndFail` to `failWhenExists`
- renamed `assertCap6PaasCreated`  to `assertCap6NoUnwantedArtifacts`

File: internal/controller/paas_controller.go (hunk 8 above)
Comment (devotional-phoenix-97) [reply internal/controller/paas_controller.go]: Yes, I reverted previous changes (silly me).
And that indeed resolved my last issue. Thanks.

File: test/e2e/capability-cap6_test.go (hunk 7 above)
Comment (devotional-phoenix-97) [reply test/e2e/capability-cap6_test.go:58]: By adding this assertion I noticed we where unable to delete a Paas without a proper config which I fixed in this PR. We shoud leave this test to make we don't degrade functionality as we change code for these edgecases.

**Review by TazmanianDevil666 — APPROVED** [review]

**Pull request merged**